### ISSUE TITLE
Problem with Morearty.Mixin & observeBinding in nested components 

### ISSUE THREAD

**Issue by honzatrtik** [moreartyjs/moreartyjs#63]
Hello @Tvaroh,
I have two components - parent & child, both of them are mixing in Morearty.Mixin, child uses this.observeBinding for triggering render, but nothing happens if binding changes. If I remove  Morearty.Mixin from parent, everything works.
Is this expected behaviour or bug?

Please, see this gist: https://gist.github.com/honzatrtik/481fc2d3bd82d28d48e8

Anyway, thank you for your awesome work on this library!

**Comment by Tvaroh** [comment]
Hello, @honzatrtik.

Thanks for the report and your kind words.

I think you've touched limitations of observed bindings that I forgot to mention in wiki. The problem is that parent knows nothing about it's children's bindings. So, if it doesn't care about some binding that is 'observed' by its child, parent's `shouldComponentUpdate` will just return `false` and the rendering will stop.

Can you see the issue?

**Comment by Tvaroh** [comment]
I've update the [wiki](https://github.com/moreartyjs/moreartyjs/wiki/Authoring-components#observed-bindings-limitations).

**Comment by honzatrtik** [comment]
Yes, your explanation makes sense. Thank you!

**Comment by Tvaroh** [comment]
Could you check this in your scenario?

**Comment by honzatrtik** [comment]
Yup, each of the ancestors must either observe all bindings as child component or didnt mix in Morearty.Mixin (provided that `shouldComponentUpdate` is not overriden).

Maybe `observeBinding` could save reference to component and invoke forceUpdate, but it kinda breaks top-down rendering philosophy...

**Comment by gilbox** [comment]
My coworker [dirtyrottenscoundrel[(https://github.com/dirtyrottenscoundrel) looked at how om implements this. It does bookkeeping on all reference bindings and does a pass over all changed ref cursors after each render of the root component completes. I think this is actually something that you could achieve with a mixin but it would be nice to work it into morearty itself.

The alternative is to observe all reference cursors (from a flux store or whatever) and force a re-render of everything any time there is a change. This seems fine if you're just using reference cursors for something like the user's logged-in status (which is the only use-case I have in mind for myself, atm). But if you use them extensively this could get expensive.

**Comment by Tvaroh** [comment]
@gilbox, interesting. How does it resolve the current limitation I described above?

**Comment by gilbox** [comment]
- There is a `ref queue`. 
- Every component with a reference binding attaches an observer to the binding. 
- Whenever the observer detects a change, it pushes something like `{ component: this, binding: binding }` into the queue. 
- Then, on the next animationFrame render(), it renders root component which renders the entire render tree.
- After render tree is completely rendered (the usual way), it iterates through `ref queue` and does 
```
          if (c.shouldComponentUpdate()) { c.forceUpdate() }
```
- Note that in om, `shouldComponentUpdate()` also checks ref bindings, not just passed bindings the way morearty does.

I suspect that any time something is added to the queue, om needs to schedule a `render()` but that part I'm not totally clear on.

**Comment by Tvaroh** [comment]
Thanks, Gil. By the way Morearty also checks all bindings, including observed ones, in `shouldComponentUpdate`. Probably we can incorporate this stuff in the near future.

I see this like so:

1. when rendering components which have observed bindings, these components should be registered by putting to some structure within context;
2. after each render Morearty should call `forceUpdate` on registered components if required (need to know which of them weren't already rendered).

**Comment by gilbox** [comment]
It seems strange for  `shouldComponentUpdate` to check observed bindings currently, since during top-down re-render it won't effect the return value of `shouldComponentUpdate` calls. But I guess you did this for when someone calls `setState` ?

**Comment by Tvaroh** [comment]
Gil, can you clarify why checking observed bindings want affect SCU?

**Comment by gilbox** [comment]
I was going to say: *because the only bindings that will effect SCU will always be a subset of the bindings in `this.props.binding`*. But I just realized that's not true: The only bindings that will effect SCU will always be a subset of the bindings in the **parent** component.

**Comment by gilbox** [comment]
A more detailed idea dump.. how ref bindings might be implemented in morearty. The idea is to modify Context and Mixin pieces of Morearty.

- Let morearty Context instance hold a hash of ref bindings (indexed by path)
- each ref binding holds a hash of components (indexed by component id)
- each component holds an array of ref bindings
- on component mount/unmount, 
    1. add each ref binding for component to Context’s ref binding hash if it doesn’t already exist there
    2. add/remove component to/from each of the component’s ref binding’s component hash
- whenever a component is removed from the components array of a ref binding, check to see if the components array is empty. if it is empty, remove that ref from the Context’s ref hash
- When ref binding is added to a component, that component observes ref binding
- Whenever ref binding changes, component schedules re-render.
- the implementation of re-render function is same as morearty's current render method, except 
    1. at the end it loops through all ref bindings and components associated with each ref binding 
    2. for each component, if #shouldComponentUpdate evaluates to true, call #forceUpdate on that component

Inefficiencies
- A single component could be re-rendered multiple times during one render pass. In Om source code I don’t see anywhere it is trying to avoid this, but I don't hardly know ClojureScript.
- Could be faster to index ref bindings by an integer instead of path

Differences 
- What I call “ref bindings” are called binding listeners (?) in morearty
- Currently Morearty stores ref bindings (aka binding listeners) in an immutable map instead of a regular javascript object. Either implementation works, but I think that the immutable implementation is slower. -Probably there is some reason to use immutable objects instead. Anyway, all of the above works the same with immutables instead.

**Comment by gilbox** [comment]
I implemented [a much simpler solution to updating components with OBs](https://github.com/gilbox/moreartyjs/commit/51f553d59d37b30924cc548452dda8422c9d5d85). I feel like there must be something wrong with it because it seems too obvious a solution, but I don't know what that might be. It is much simpler than the Om method I described above.

Here is the gist of it:

- Add a property to Context `_componentQueue` which is a plain JS object
- Add a method to Context `#addComponentToRenderQueue(component)` which will add a component to the `_componentQueue`, keyed by it's unique component id. This way, it prevents the same component from being added more than 1X.
- Whenever an observed binding is added to a component, do `binding.addListener` and inside of the listener call `#addComponentToRenderQueue(self)`
- Modify the morearty re-render function to loop through `_componentQueue` and call `#forceUpdate` on all components.

[This test](https://github.com/gilbox/moreartyjs/commit/51f553d59d37b30924cc548452dda8422c9d5d85#diff-3a6ea288fd84ec582eb20c548f595402R1229) proves that it works. However, these changes break other tests which assume that observed bindings don't have the enhanced behavior so if you try this out, you should just run this single test using the command:

```
./node_modules/mocha/bin/mocha -b -R spec test/* -g 'when observed binding changed'
```

Obviously, other tests will need to be updated if this is to be adopted but since I'm not sure if this approach will be acceptable, this is still just a proof of concept.

I also created a [second test](https://github.com/gilbox/moreartyjs/commit/51f553d59d37b30924cc548452dda8422c9d5d85#diff-3a6ea288fd84ec582eb20c548f595402R1271), which should fail if this implementation causes double-render of a component but to my surprise it did not fail. I think that after first render SCU returns `false`, and because I render the components with observed bindings first (unlike Om which does it second), a double-render is avoided.

@Tvaroh or anyone else, any thoughts on whether this is a good approach?

**Comment by Tvaroh** [comment]
@gilbox, sounds nice. One question: does this prevent double-or-more-render when components A -> B -> C -> D from the same hierarchy use observed bindings? Won't it call `forceUpdate` on A, then on B, and so on, causing component D to re-render 4 times?

**Comment by gilbox** [comment]
Good point @Tvaroh I think that the way to deal with this is call SCU before `#forceUpdate` ([Om does this](https://github.com/omcljs/om/blob/master/src/om/core.cljs#L1157)) ... but I don't really understand how to ensure that subsequent calls to SCU can return false. Ie, two subsequent calls to SCU currently both return `true`.

**Comment by Tvaroh** [comment]
Gil, how do React's component ids look like?

**Comment by gilbox** [comment]
something like `.j20dj309j3` but they are not unique per component, only per dom node that's why I had to create my [own id generator](https://github.com/gilbox/moreartyjs/commit/0db3304510857eb43c1361cc782002f0c1465d33#diff-964ba8158b899e0a296ee0ce9dbc7d85R453)

**Comment by gilbox** [comment]
It seems that the following methods are returning `true` even after render. Any idea how to fix this?

```
bindingChanged = function (context, currentBinding) {
  return (context._stateChanged && currentBinding.isChanged(context._previousState)) ||
         (context._metaChanged && context._metaBinding.sub(currentBinding.getPath()).isChanged(context._previousMetaState));
};

observedBindingsChanged = function (self) {
  return self.observedBindings &&
      !!Util.find(self.observedBindings, function (binding) {
        return bindingChanged(self.getMoreartyContext(), binding);
      });
};
```

**Comment by gilbox** [comment]
[Another idea](https://github.com/gilbox/moreartyjs/commit/3305cf252416d8eefaab32d04fdb5cd97be596d8)... 

I haven't considered all edge cases, but I am relying on parent element always having lower index in `_componentQueue` than child. When this is the case, we can

               delete ctx._componentQueue[this.componentQueueId];

in SCU to prevent double-render. Child removes itself from the queue before child's `#forceRender` fires as a result of observed binding.

**Comment by Tvaroh** [comment]
@gilbox, could you detail the algorithm in steps?

**Comment by gilbox** [comment]
- In `#componentWillMount` every component is assigned a unique *component id*, ie., index via `Context#getUniqueComponentQueueId`. The app's `Context` instance keeps track of the last index. It should be guaranteed that a child id is always higher than a parent id, this becomes important later on.

- Every time an observed binding is registered (via Morearty's `#observeBinding`) inside of a component, we create an event listener that detects when the observed binding changes. When that observed binding changes, add the component to a **OB render queue** which is indexed by it's component id ( `Context#addComponentToRenderQueue(<component>)` ). 

- In SCU, a component always removes itself from the **OB render queue**.

- Note that the "OB render queue" is not a normal queue because the order of rendering will be determined by component id, *not* by order it was added to the queue. This allows us to prevent double-render

- Note that whenever an observed binding changes, a top-down render is queued. My implementation does not try to force this render because it assumes morearty will always perform a top-down render when any binding changes.

- Now when top-down render is triggered, before morearty calls `rootComp.forceUpdate()`, we do:
   * Iterate through the **OB render queue** and call forceUpdate() on all components in the queue
   * Note that we iterate in order of id
   * Note that double-render is avoided because a child component will remove itself from the queue if it so happens that it's parent `render()` causes it to re-render.

**Comment by Tvaroh** [comment]
Gil, can some component be double rendered first from OB processing stage and then from general `rootComp.forceUpdate()`?

**Comment by gilbox** [comment]
In my experiments that is not the case. Ie.,

`SCU` => `render()` => `SCU`  ----> `true` => `render()` => `false`

For this reason, anything that gets forceUpdate during OB processing stage will not be rendered as a result of `rootComp.forceUpdate()`

I believe that the UT "should always re-render when observed binding changed" proves this correct.

**Comment by Tvaroh** [comment]
Cool, could you review the code and make pull request? Then I will also analyze it for corner cases.

**Comment by gilbox** [comment]
I realized that my last response maybe contradicted what I said earlier. I'll experiment with some more tests to see if I can get the algorithm to double-render.

**Comment by gilbox** [comment]
I was able to modify a test to cause double-render in the case you described. I fixed it with [additional bookkeeping](https://github.com/gilbox/moreartyjs/commit/edb8359140d7b253fc84c2554b1258d709adaa59).

Btw, I am running tests on react .12 because with .13 I get errors related to test bootstrapping.

**Comment by Tvaroh** [comment]
Hm, current tests work okay with React 0.13. Do you have something specific in your tests? Maybe try to remove *node_modules* and `npm install` from scratch?

**Comment by gilbox** [comment]
Do you mean you get these errors but it's okay because the tests still pass?

```
Error details: Invariant Violation: dangerouslyReplaceNodeWithMarkup(...): Cannot render markup in a worker thread. Make sure `window` and `document` are available globally before requiring React when unit testing or use React.renderToString for server rendering. Error: Invariant Violation: dangerouslyReplaceNodeWithMarkup(...): Cannot render markup in a worker thread. Make sure `window` and `document` are available globally before requiring React when unit testing or use React.renderToString for server rendering.
```

**Comment by Tvaroh** [comment]
Yep, but that error appears IIRC in single test. Am I wrong?

**Comment by Tvaroh** [comment]
Hm, I am wrong. With React 0.13 it appears many times.

**Comment by gilbox** [comment]
I think that the problem is that React is loading before `global.window` is created.

**Comment by gilbox** [comment]
Here's my branch fixing those UT errors. However, there are still some other non-related errors. 

This one is troubling although the test runner still reports that all tests pass:

```
Morearty: render error. Continuing.
Error details: expected 'foo' to equal 'value2' AssertionError: expected 'foo' to equal 'value2'
    at Function.assert.strictEqual (/Users/gbirman/iz/moreartyjs/node_modules/chai/lib/chai/interface/assert.js:169:32)
    at /Users/gbirman/iz/moreartyjs/test/Morearty.js:1240:22
    at module.exports.Mixin.observeBinding (/Users/gbirman/iz/moreartyjs/src/Morearty.js:566:19)
    at createClass.render (/Users/gbirman/iz/moreartyjs/test/Morearty.js:1238:25)
    at ReactCompositeComponentMixin._renderValidatedComponentWithoutOwnerOrContext (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:767:34)
    at ReactCompositeComponentMixin._renderValidatedComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:793:14)
    at wrapper [as _renderValidatedComponent] (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactCompositeComponentMixin._updateRenderedComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:724:36)
    at ReactCompositeComponentMixin._performComponentUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:705:10)
    at ReactCompositeComponentMixin.updateComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:625:12)
    at wrapper [as updateComponent] (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactCompositeComponentMixin.receiveComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:489:10)
    at Object.ReactReconciler.receiveComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactReconciler.js:97:22)
    at ReactCompositeComponentMixin._updateRenderedComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:726:23)
    at ReactCompositeComponentMixin._performComponentUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:705:10)
    at ReactCompositeComponentMixin.updateComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:625:12)
    at wrapper [as updateComponent] (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactCompositeComponentMixin.receiveComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:489:10)
    at Object.ReactReconciler.receiveComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactReconciler.js:97:22)
    at ReactCompositeComponentMixin._updateRenderedComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:726:23)
    at ReactCompositeComponentMixin._performComponentUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:705:10)
    at ReactCompositeComponentMixin.updateComponent (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:625:12)
    at wrapper [as updateComponent] (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactCompositeComponentMixin.performUpdateIfNecessary (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactCompositeComponent.js:522:12)
    at Object.ReactReconciler.performUpdateIfNecessary (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactReconciler.js:115:22)
    at runBatchedUpdates (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdates.js:151:21)
    at ReactReconcileTransaction.Mixin.perform (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/Transaction.js:134:20)
    at ReactUpdatesFlushTransaction.Mixin.perform (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/Transaction.js:134:20)
    at ReactUpdatesFlushTransaction.assign.perform (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdates.js:95:38)
    at Object.flushBatchedUpdates (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdates.js:175:19)
    at Object.wrapper [as flushBatchedUpdates] (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/Transaction.js:207:25)
    at ReactDefaultBatchingStrategyTransaction.Mixin.perform (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/Transaction.js:148:16)
    at Object.ReactDefaultBatchingStrategy.batchedUpdates (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactDefaultBatchingStrategy.js:66:19)
    at Object.enqueueUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdates.js:215:22)
    at enqueueUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdateQueue.js:30:18)
    at Object.ReactUpdateQueue.enqueueForceUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactUpdateQueue.js:153:5)
    at ReactComponent.forceUpdate (/Users/gbirman/iz/moreartyjs/node_modules/react/lib/ReactComponent.js:90:20)
    at /Users/gbirman/iz/moreartyjs/src/Morearty.js:390:20
    at catchingRenderErrors (/Users/gbirman/iz/moreartyjs/src/Morearty.js:365:11)
    at render [as _onTimeout] (/Users/gbirman/iz/moreartyjs/src/Morearty.js:382:7)
    at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
```

**Comment by Tvaroh** [comment]
Cool! I will look into that assertion error tomorrow.

**Comment by gilbox** [comment]
Question -- 

```
bindingChanged = function (context, currentBinding) {
  return (context._stateChanged && currentBinding.isChanged(context._previousState)) ||
         (context._metaChanged && context._metaBinding.sub(currentBinding.getPath()).isChanged(context._previousMetaState));
};
```

Instead of checking context._previousState, why not compare the previous and current binding's reference to Immutable object ?

**Comment by Tvaroh** [comment]
Gil, I ported your tests fixes into master and removed that wrong assertion. Regarding your last question - I can't answer immediately, need to check.

**Comment by gilbox** [comment]
Re: gilbox-ob-queue-simplified ... If a user defines `shouldComponentUpdateOverride`, it's possible that component._previousState will not really represent the previous state. For example:

    // define shouldComponentUpdateOverride:
    SCUO = (SCU) => return SCU() && this._meaninglessIndex === 0;
    
    // initial value of binding passed into the component
    // binding.get() === 'foo'
    binding.set('one')
    // component._previousState.get() === 'foo'
    // => SCUO returns false because component._meaninglessIndex === 0
    // component._previousState.get() === 'foo'
    binding.set('two')
    // component._previousState.get() === 'foo'
    // => SCUO returns true because binding value changes and component._meaninglessIndex !== 0
    // component._previousState.get() === 'two'

It's pretty easy to fix this, but I'm not sure if it matters. What do you think?

**Comment by Tvaroh** [comment]
I think less surprises is better, so let's incorporate this.

**Comment by gilbox** [comment]
I think this is ready.

**Comment by Tvaroh** [comment]
Published as part of *0.7.15*. All credits to Gil!
